src/rideables/TGraph.hpp

```diff
@@ -161,171 +258,240 @@ class TGraph : public RGraph{
+                        for (auto r : destination(other)) {
+                            if (r->src == vid) {
+                                std::cout << "Them: (" << r->src << "," << r->dest << ")" << std::endl;
+                            }
                         }
+                        for (auto r : destination(vid)) {
+                            if (r->src == other) {
+                                std::cout << "Us: (" << r->src << "," << r->dest << ")" << std::endl;
+                            }
+                        }
+                        for (auto r : source(other)) {
+                            if (r->dest == vid) {
+                                std::cout << "Them: (" << r->src << "," << r->dest << ")" << std::endl;
+                            }
+                        }
+                        std::abort();
                     }
-                    for (Relation *r : toRemoveList) {
-                        _v->dest_list.erase(r);
-                        garbageList.push_back(r);
-                    }
-                }
+                    remove_relation(source(other), &src);
+                    remove_relation(destination(other), &dest);
+                    assert(!has_relation(source(other), &src) && !has_relation(destination(other), &dest));
+                }                
                 
-                // Step 4: Delete edges, clear set of src and dest edges
-                v->adjacency_list.clear();
-                v->dest_list.clear();
-                for (Relation *r : garbageList) {
-                    delete r;
-                }
+                std::vector<Relation*> toDelete(source(vid).size() + destination(vid).size());
+                for (auto r : source(vid)) toDelete.push_back(r);
+                for (auto r : destination(vid)) toDelete.push_back(r);
+                source(vid).clear();
+                destination(vid).clear();
+                destroy(vid);
+                for (auto r : toDelete) delete r;
                 
-                // Step 5: Release in reverse order
+                // Step 4: Release in reverse order
                 std::reverse(vertices.begin(), vertices.end());
                 for (int _vid : vertices) {
-                    idxToVertex[_vid]->seqNumber++;
-                    idxToVertex[_vid]->unlock();
+                    inc_seq(_vid);
+                    unlock(_vid);
                 }
             }
             return true;
         }
         
-        void for_each_edge(int v, std::function<bool(int)> fn) {
-            idxToVertex[v]->lock();
-            for (Relation *r : idxToVertex[v]->adjacency_list) {
-                if (!fn(r->dest)) {
-                    break;
-                }
+    private:
+        void lock(size_t idx) {
+            std::atomic<bool>& lck = vertexLocks[idx];
+            bool expect = false;
+            while (lck.load() == true || lck.compare_exchange_strong(expect, true) == false) {
```

Comment:
`std::atomic<bool>` has a size of 1 byte, and `std::mutex` has a size of 40 bytes; we maintain one lock per vertex, so shrinking the space requirement by 40x is ideal here, especially since the locks are held in transient memory.

Reply:
I’m worried about the overhead of being busy-waiting. Also, 1 Byte suffers from false sharing. Honestly, 40B*1M (Orkut has ~3M vertices which is close to our 1M) doesn’t sound big to me, especially given that we already use 4B as sequence number and 8B as idxToVertex indexing for each vertex, let alone transient tVertex which is at least 40B; I don’t think the size of mutex would become the space bottleneck, but it’s possible to be the performance bottleneck.